test/integration/smoke/test_async_job.py

```diff
@@ -0,0 +1,160 @@
+            cls.api_client,
+            cls.testdata["disk_offering"]
+        )
+
+        cls._cleanup = [
+            cls.service_offering,
+            cls.disk_offering
+        ]
+        return
+
+    @classmethod
+    def tearDownClass(cls):
+        try:
+            cleanup_resources(cls.api_client, cls._cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+
+    def setUp(self):
+        self.apiclient = self.testClient.getApiClient()
+        self.dbclient = self.testClient.getDbConnection()
+        self.hypervisor = self.testClient.getHypervisorInfo()
+        self.testdata["virtual_machine"]["zoneid"] = self.zone.id
+        self.testdata["virtual_machine"]["template"] = self.template.id
+        self.testdata["iso"]["zoneid"] = self.zone.id
+        self.account = Account.create(
+            self.apiclient,
+            self.testdata["account"],
+            domainid=self.domain.id
+        )
+        self.cleanup = [self.account]
+        return
+
+    def tearDown(self):
+        try:
+            self.debug("Cleaning up the resources")
+            cleanup_resources(self.apiclient, self.cleanup)
+            self.debug("Cleanup complete!")
+        except Exception as e:
+            self.debug("Warning! Exception in tearDown: %s" % e)
+
+    @attr(
+        tags=[
+            "advanced",
+            "eip",
+            "advancedns",
+            "basic",
+            "sg"],
+        required_hardware="true")
+    def test_queryAsyncJobResult(self):
+        """
+        Test queryAsyncJobResult API for expected values
+        """
+        self.debug("Deploying instance in the account: %s" %
+                   self.account.name)
+        self.virtual_machine = VirtualMachine.create(
+            self.apiclient,
+            self.testdata["virtual_machine"],
+            accountid=self.account.name,
+            domainid=self.account.domainid,
+            serviceofferingid=self.service_offering.id,
+            diskofferingid=self.disk_offering.id,
+            hypervisor=self.hypervisor
+        )
+
+        response = self.virtual_machine.getState(
+            self.apiclient,
+            VirtualMachine.RUNNING)
+        self.assertEqual(response[0], PASS, response[1])
+
+        cmd = queryAsyncJobResult.queryAsyncJobResultCmd()
+        cmd.jobid = self.virtual_machine.jobid
+        cmd_response = self.apiclient.queryAsyncJobResult(cmd)
+        endtime = cmd_response.endtime
+        self.assertIsNotNone(endtime, "Expected 'endtime' field of queryAsyncJobResult to have a timestamp of when the job finished.")
+        return
+
+    @attr(
+        tags=[
+            "advanced",
+            "eip",
+            "advancedns",
+            "basic",
+            "sg"],
+        required_hardware="true")
+    def test_async_job_table_values(self):
+        """
+        Test async job db table for valid values
+        """
```

Comment:
Add another failure test case. Destroy something twice, etc. job status -> failed (int 2?).

Reply:
I've tried this case and I'm not sure it baheves as expected. I've successfully destroyed a VM and tried to do it again and the job result was successful same as the first try. Should this behave like that? I think it should return an error saying the VM is already in Destroyed state.

Reply:
I get the same result as you @borisstoyanov

Reply:
@ernjvr do you think we should refactor the service if it tries to bring a VM into a state which it is, to prompt an error? It kinda makes sense to me.

Reply:
@borisstoyanov I investigated the code for the scenario of starting a VM. It also returns a success status when I make a request to start a VM that has already been started. This success status is returned by the VM job handler code. The async job status gets updated with whatever status is returned by the VM job handler code. So, I think changing how the VM handler code deals with duplicate commands for the same resource is out of scope for this feature. I think it should be dealt with in a separate bug fix ticket. What do you think?

Reply:
Sound the right way to do it, thanks for the investigation @ernjvr, let me log a ticket in community. When we have time we could get back to this.

Reply:
https://github.com/apache/cloudstack/issues/2736